dotnet/src/SemanticKernel.Abstractions/AI/FunctionChoiceBehaviors/AutoFunctionChoiceBehavior.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Represent <see cref="FunctionChoiceBehavior"/> that provides either all of the <see cref="Kernel"/>'s plugins' function information to the model or a specified subset.
+/// This behavior allows the model to decide whether to call the functions and, if so, which ones to call.
+/// </summary>
+[Experimental("SKEXP0001")]
+public sealed class AutoFunctionChoiceBehavior : FunctionChoiceBehavior
+{
+    /// <summary>
+    /// List of the functions that the model can choose from.
+    /// </summary>
+    private readonly IEnumerable<KernelFunction>? _functions;
+
+    /// <summary>
+    /// List of the fully qualified names of the functions that the model can choose from.
+    /// </summary>
+    private readonly IEnumerable<string>? _functionFQNs;
+
+    /// <summary>
+    /// This class type discriminator used for polymorphic deserialization of the type specified in JSON and YAML prompts.
+    /// </summary>
+    public const string TypeDiscriminator = "auto";
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AutoFunctionChoiceBehavior"/> class.
+    /// </summary>
+    [JsonConstructor]
+    public AutoFunctionChoiceBehavior()
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AutoFunctionChoiceBehavior"/> class.
+    /// </summary>
+    /// <param name="functions">The subset of the <see cref="Kernel"/>'s plugins' functions information.</param>
+    public AutoFunctionChoiceBehavior(IEnumerable<KernelFunction> functions)
+    {
+        this._functions = functions;
+    }
+
+    /// <summary>
+    /// The maximum number of function auto-invokes that can be made in a single user request.
+    /// </summary>
+    /// <remarks>
+    /// After this number of iterations as part of a single user request is reached, auto-invocation
+    /// will be disabled. This is a safeguard against possible runaway execution if the model routinely re-requests
+    /// the same function over and over. To disable auto invocation, this can be set to 0.
+    /// </remarks>
+    [JsonPropertyName("maximumAutoInvokeAttempts")]
+    public int MaximumAutoInvokeAttempts { get; init; } = DefaultMaximumAutoInvokeAttempts;
+
+    /// <summary>
+    /// Fully qualified names of subset of the <see cref="Kernel"/>'s plugins' functions information to provide to the model.
+    /// </summary>
+    [JsonPropertyName("functions")]
+    public IEnumerable<string>? Functions
+    {
+        get => this._functionFQNs;
+        init
+        {
+            if (value?.Count() > 0 && this._functions?.Count() > 0)
+            {
+                throw new KernelException("Functions are already provided via the constructor.");
```

Comment:
Can property just override what's in constructor? It's worth to provide some comments in code about this logic.

Reply:
Today, the existing function-calling functionality supports **manual** invocation of kernel functions created via `KernelFunctionFactory.CreateFrom*` but non-imported into the kernel. Auto function invocation requires functions to be registered in the kernel, though.  
   
The new model partially follows the existing one and also supports non-imported into kernel functions for manual invocation. To register these non-imported functions, their **instances** should be passed as an argument to one of the choice behavior class constructors.  
   
Additionally, the new model supports kernel functions specified via prompts as a list of fully qualified function names - ["plugin1.function1", "plugin1.function2"]. The function list is assigned to the `Function` property during the deserialization process.  
   
Taking that context into account, there are a few options to consider:  
1. Stop supporting non-imported kernel functions. This way, all the functions passed to the new model class constructors will be converted to the list of fully qualified function names and assigned to the `Functions` property. The `GetConfiguration` method will verify if all the functions specified in the `Functions` property are imported into the kernel and throw an exception if any is non imported.  
2. Give preference to functions specified in the `Functions` property over those passed via the constructor. The `GetConfiguration` method will always use functions from the `Functions` property if available, and use the ones passed via the constructor otherwise.
3. Throw an exception if callers pass functions via constructor and via the `Functions` property. This is the implementation in this PR.

@stephentoub, are you aware of any use case that requires support for non-imported functions in the kernel?

Reply:
Fixed, by initializing the property from functions provided via the class constructor.